pkg/limits/limits.go

```diff
@@ -17,4 +22,68 @@ const (
 	TTLMax                = 30 * time.Minute
 	ClusterAffiliatesMax  = 1024
 	AffiliateEndpointsMax = 64
+	RequestsPerSecondMax  = 5
 )
+
+// IPRateLimiter type.
+type IPRateLimiter struct {
+	IPs        map[string]*rate.Limiter
+	Mu         *sync.RWMutex
+	RateLimit  rate.Limit
+	BucketSize int
+}
+
+// NewIPRateLimiter returns newly initialized IPRateLimiter.
+func NewIPRateLimiter(rateLimit rate.Limit, bucketSize int) *IPRateLimiter {
+	i := &IPRateLimiter{
+		IPs:        make(map[string]*rate.Limiter),
+		Mu:         &sync.RWMutex{},
+		RateLimit:  rateLimit,
+		BucketSize: bucketSize,
+	}
+
+	return i
+}
+
+// AddIP creates a new rate limiter and adds it to the ips map,
+// using the IP address as the key.
+func (iPRL *IPRateLimiter) AddIP(ip string) *rate.Limiter {
+	iPRL.Mu.Lock()
+	defer iPRL.Mu.Unlock()
+
+	limiter := rate.NewLimiter(iPRL.RateLimit, iPRL.BucketSize)
+
+	iPRL.IPs[ip] = limiter
+
+	return limiter
+}
+
+// GetLimiter returns the rate limiter for the provided IP address if it exists.
+// Otherwise calls AddIP to add IP address to the map.
+func (iPRL *IPRateLimiter) GetLimiter(ip string) *rate.Limiter {
+	iPRL.Mu.Lock()
+	limiter, exists := iPRL.IPs[ip]
+
+	if !exists {
+		iPRL.Mu.Unlock()
+
+		return iPRL.AddIP(ip)
```

Comment:
this is racy, as if two calls go out for `GetLimiter()`, they might both see limiter missing and both create new limiter via `AddIP()`.

I think it's better to unify `Add` and `Get` into a single method `Get` which creates the limiter without releasing the lock if that is missing.